### PR TITLE
867

### LINKED ISSUE #867: Map legends should be generated by the data fetching component and aggregated to the top level.
Currently map legends are set by the Map Composer. So that the legends appear in template maps etc. they should be inserted by the component that fetches them.

### PR DESCRIPTION
**Pull request by leehart**
Fixes #867 

A working solution, but a bit rough around the edges.

### CODE REVIEW

File: webapp/src/js/components/Map/TableMarkersLayer.js [hunk 1]
```diff
@@ -82,6 +88,37 @@ let TableMarkersLayer = React.createClass({
...
+      let legendControl = this.composeLegendControl({table, markerColourProperty});
+      legendControl.id = this.legendControlId;
+      changeControl(legendControl);
+    }
+  },
+
+  componentWillUnmount() {
+    let {removeControl} = this.context;
+
+    if (removeControl !== undefined) {
```
Comment (benjeffery) [inline webapp/src/js/components/Map/TableMarkersLayer.js]: Are there situations where you expect the context to not be available?

File: webapp/src/js/components/Map/TableMarkersLayer.js [hunk 2]
```diff
@@ -282,6 +319,27 @@ let TableMarkersLayer = React.createClass({
       });
   },
 
+  composeLegendControl(payload) {
+    let {table, markerColourProperty} = payload;
+
+    let legendControl = null;
+
+    if (table !== undefined && markerColourProperty !== undefined) {
+      let legend = ReactDOMServer.renderToStaticMarkup(
```
Comment (benjeffery) [inline webapp/src/js/components/Map/TableMarkersLayer.js]: Might be cleaner if the `renderToStaticMarkup` happened on the other side of the add/change call in `Map.js`?

File: webapp/src/js/components/Map/TableMarkersLayer.js (hunk 1 above)
Comment (leehart) [inline webapp/src/js/components/Map/TableMarkersLayer.js]: I expect that's a superfluous relic from when removeControl wasn't defined by Map's getChildContext(); I'll trim it out.

File: webapp/src/js/components/Map/TableMarkersLayer.js (hunk 2 above)
Comment (leehart) [inline webapp/src/js/components/Map/TableMarkersLayer.js]: I see, you mean effectively return `{component: <PropertyLegend [...] />, [...]}` and then do something like `div.innerHTML = ReactDOMServer.renderToStaticMarkup(control.component)` in Map instead? OK, will do.

**Comment by leehart** [comment]
Shall we scribble your `LeafletLegend` ideas on the kitchen wall? Any excuse! :-D

**Comment by benjeffery** [comment]
YES! We shall stake a claim!

**Pull request closed without merging**